Mathlib/Tactic/ProdAssoc.lean

```diff
@@ -0,0 +1,150 @@
+
+/-- Make a `ProdTree` out of an `Expr`. -/
+def mkProdTree (e : Expr) : MetaM ProdTree :=
+  match e with
+    | .app (.app (.const ``Prod [u,v]) X) Y => do
+        return .prod (← X.mkProdTree) (← Y.mkProdTree) u v
+    | X => do
+      let some u := (← inferType X).type? | throwError "Not a type{indentExpr X}"
+      return .type X u
+
+/-- Given `P : ProdTree` representing an iterated product and `e : Expr` which
+should correspond to a term of the iterated product, this will return
+a list, whose items correspond to the leaves of `P` (i.e. the types appearing in the product),
+where each item is the appropriate composition of `Prod.fst` and `Prod.snd` applied to `e`
+resulting in an element of the type corresponding to the leaf.
+
+For example, if `P` corresponds to `(X × Y) × Z` and `t : (X × Y) × Z`, then this
+should return `[t.fst.fst, t.fst.snd, t.snd]`.
+-/
+def ProdTree.unpack (t : Expr) : ProdTree → MetaM (List Expr)
+  | type _ _ => return [t]
+  | prod fst snd u v => do
+      let fst' ← fst.unpack <| mkAppN (.const ``Prod.fst [u,v]) #[fst.getType, snd.getType, t]
+      let snd' ← snd.unpack <| mkAppN (.const ``Prod.snd [u,v]) #[fst.getType, snd.getType, t]
+      return fst' ++ snd'
+
+/-- This function should act as the "reverse" of `ProdTree.unpack`, constructing
+a term of the iterated product out of a list of terms of the types appearing in the product. -/
+def ProdTree.pack (ts : List Expr) : ProdTree → MetaM Expr
+  | type _ _ => do
+    match ts with
+      | [] => throwError "Can't pack the empty list."
+      | [a] => return a
+      | _ => throwError "Failed due to size mismatch."
+  | prod fst snd u v => do
+    let fstSize := fst.size
+    let sndSize := snd.size
+    unless ts.length == fstSize + sndSize do throwError "Failed due to size mismatch."
+    let tsfst := ts.toArray[:fstSize] |>.toArray.toList
+    let tssnd := ts.toArray[fstSize:] |>.toArray.toList
+    let mk : Expr := mkAppN (.const ``Prod.mk [u,v]) #[fst.getType, snd.getType]
+    return .app (.app mk (← fst.pack tsfst)) (← snd.pack tssnd)
+
+/-- Converts a term `e` in an iterated product `P1` into a term of an iterated product `P2`.
+Here `e` is an `Expr` representing the term, and the iterated products are represented
+by terms of `ProdTree`. -/
+def ProdTree.convertTo (P1 P2 : ProdTree) (e : Expr) : MetaM Expr :=
+  return ← P2.pack <| ← P1.unpack e
+
+/-- Given two expressions corresponding to iterated products of the same types, associated in
+possibly different ways, this constructs the "obvious" function from one to the other. -/
+def mkProdFun (a b : Expr) : MetaM Expr := do
+  let pa ← a.mkProdTree
+  let pb ← b.mkProdTree
+  unless pa.components.length == pb.components.length do
+    throwError "The number of components of {a} and {b} must match."
+  for (x,y) in pa.components.zip pb.components do
+    unless ← isDefEq x y do
+      throwError "Components {x} is not defeq to component {y}."
```

Comment:
Small typo, and I'd suggest writing the message out like this:
```suggestion
      throwError "Component{indentD x}\nis not definitionally equal to component{indentD y}."
```

Reply:
Done. Thanks!